src/mito2/src/read/dedup.rs

```diff
@@ -211,11 +223,258 @@ pub(crate) struct DedupMetrics {
     pub(crate) num_deleted_rows: usize,
 }
 
+/// Buffer to store fields in the last row to merge.
+struct LastFieldsBuilder {
+    /// Filter deleted rows.
+    filter_deleted: bool,
+    /// Fields builders, lazy initialized.
+    builders: Vec<Box<dyn MutableVector>>,
+    /// Last fields to merge, lazy initialized.
+    last_fields: Vec<Value>,
+    /// Whether the last row (including `last_fields`) has null field.
+    has_null: bool,
+    /// Whether the last row has delete op. If true, skips merging fields.
+    has_delete: bool,
+    /// Whether the builder is initialized.
+    initialized: bool,
+}
+
+impl LastFieldsBuilder {
+    /// Returns a new builder with the given `filter_deleted` flag.
+    fn new(filter_deleted: bool) -> Self {
+        Self {
+            filter_deleted,
+            builders: Vec::new(),
+            last_fields: Vec::new(),
+            has_null: false,
+            has_delete: false,
+            initialized: false,
+        }
+    }
+
+    /// Initializes the builders with the last row of the batch.
+    fn maybe_init(&mut self, batch: &Batch) {
+        debug_assert!(!batch.is_empty());
+
+        if self.initialized || batch.fields().is_empty() {
+            // Already initialized or no fields to merge.
+            return;
+        }
+
+        self.initialized = true;
+
+        let last_idx = batch.num_rows() - 1;
+        let fields = batch.fields();
+        self.has_null = fields.iter().any(|col| col.data.is_null(last_idx));
```

Comment:
When has_null == false, can we return early without calculating has_delete?

Reply:
Maybe we can compute `has_delete` first and skip computing has_null if it is true.